diff --git a/squirrel/sqbaselib.cpp b/squirrel/sqbaselib.cpp
index ae54d39..fed42d9 100644
--- a/squirrel/sqbaselib.cpp
+++ b/squirrel/sqbaselib.cpp
@@ -1854,6 +1854,7 @@ const SQRegFunction SQSharedState::_string_default_delegate_funcz[]={
     {_SC("hash"),string_hash, 1, _SC("s")},
     {_SC("slice"),string_slice,-1, _SC("s n  n")},
     {_SC("indexof"),string_indexof,-2, _SC("s s n")},
+    {_SC("find"),string_indexof,-2, _SC("s s n")},
     {_SC("contains"),string_contains,-2, _SC("s s n")},
     {_SC("tolower"),string_tolower,-1, _SC("s n n")},
     {_SC("toupper"),string_toupper,-1, _SC("s n n")},
